src/hmem_cuda.c

```diff
@@ -73,50 +125,19 @@ struct cuda_ops {
 	cudaError_t (*cudaIpcGetMemHandle)(cudaIpcMemHandle_t *handle,
 					   void *devptr);
 	cudaError_t (*cudaIpcCloseMemHandle)(void *devptr);
-};
-
-static bool hmem_cuda_use_gdrcopy;
-static enum cuda_xfer_setting hmem_cuda_xfer_setting;
-static bool cuda_ipc_enabled;
-static int cuda_device_count;
+} cuda_ops
+#if !ENABLE_CUDA_DLOPEN
+#define CUDA_OPS_INIT(sym) .sym = sym,
+= {
+	CUDA_DRIVER_FUNCS_DEF(CUDA_OPS_INIT)
+	CUDA_RUNTIME_FUNCS_DEF(CUDA_OPS_INIT)
+}
+#endif
```

Comment:
Oooooof. Stumped on this for a second.

For readability I think it's better to split declaration and initialization, e.g.
```
static struct {
	cudaError_t (*cudaMemcpy)(void *dst, const void *src, size_t size,
	cudaError_t (*cudaMemcpy)(void *dst, const void *src, size_t size,
				  enum cudaMemcpyKind kind);
				  enum cudaMemcpyKind kind);
	cudaError_t (*cudaDeviceSynchronize)(void);
	cudaError_t (*cudaDeviceSynchronize)(void);
@@ -61,6 +111,8 @@ struct cuda_ops {
					  CUdeviceptr ptr);
					  CUdeviceptr ptr);
	CUresult (*cuMemGetAddressRange)( CUdeviceptr* pbase,
	CUresult (*cuMemGetAddressRange)( CUdeviceptr* pbase,
					  size_t* psize, CUdeviceptr dptr);
					  size_t* psize, CUdeviceptr dptr);
	CUresult (*cuDeviceGetP2PAttribute)(int* value, CUdevice_P2PAttribute
			attrib, CUdevice srcDevice, CUdevice dstDevice);
	cudaError_t (*cudaHostRegister)(void *ptr, size_t size,
	cudaError_t (*cudaHostRegister)(void *ptr, size_t size,
					unsigned int flags);
					unsigned int flags);
	cudaError_t (*cudaHostUnregister)(void *ptr);
	cudaError_t (*cudaHostUnregister)(void *ptr);
@@ -73,50 +125,19 @@ struct cuda_ops {
	cudaError_t (*cudaIpcGetMemHandle)(cudaIpcMemHandle_t *handle,
	cudaError_t (*cudaIpcGetMemHandle)(cudaIpcMemHandle_t *handle,
					   void *devptr);
					   void *devptr);
	cudaError_t (*cudaIpcCloseMemHandle)(void *devptr);
	cudaError_t (*cudaIpcCloseMemHandle)(void *devptr);
};
} cuda_ops;

#if !ENABLE_CUDA_DLOPEN
cuda_ops = ...
#endif
```

Reply:
That wouldn't work as a drop-in replacement without the struct being defined/`typedef`'d before declaring `cuda_ops`. I intentionally converted this into an untyped struct literal since `cuda_ops` is a singleton object.